### PR TITLE
refactor our coordinator design pattern to depend on protocols not enums

### PR DESCRIPTION
**Pull request by mahmoudmorsy201**

### CODE REVIEW

File: Mercury-Store/Mercury-Store/Support/Coordinator/Coordinator.swift [hunk 1]
```diff
@@ -7,27 +7,26 @@
 
 import UIKit
 
-protocol FlowCoordinator: AnyObject {
-    var parentCoordinator: MainBaseCoordinator? {get set}
-}
+import Foundation
+import UIKit
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Support/Coordinator/Coordinator.swift:11]: I see duplicate imports

File: Mercury-Store/Mercury-Store/Profile/Views/LoginView/LoginViewController.swift [hunk 2]
```diff
@@ -67,10 +54,10 @@ class LoginViewController: UIViewController,ProfileCoordinated {
     @IBAction func loginPressed(_ sender: Any) {
         
         print("Tapped Login Button")
-        coordinator?.moveTo(flow: .profile(.intialScreen), userData: nil)
+        //coordinator?.moveTo(flow: .profile(.intialScreen), userData: nil)
     }
     @IBAction func goToSignUp(_ sender: UIButton) {
-        coordinator?.moveTo(flow: .profile(.registerScreen), userData: nil)
+        //coordinator?.moveTo(flow: .profile(.registerScreen), userData: nil)
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Profile/Views/LoginView/LoginViewController.swift:60]: Please add a flow or remove lines entirely

File: Mercury-Store/Mercury-Store/Profile/ViewModels/ProfileViewModel.swift [hunk 3]
```diff
@@ -8,11 +8,22 @@
...
+    func goToMyWishListScreen()
+    func goToMyAddressesScreen()
+    func goToAboutUsScreen()
+    func goToMainTab()
+}
+
 class ProfileViewModel {
-    
-    
+    weak var profileNavigationFlow: ProfileNavigationFlow!
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Profile/ViewModels/ProfileViewModel.swift:20]:
```suggestion
    private weak var profileNavigationFlow: ProfileNavigationFlow!
```

File: Mercury-Store/Mercury-Store/Home/ViewModels/HomeViewModel.swift [hunk 4]
```diff
@@ -6,8 +6,15 @@
 //
 
 import RxSwift
+protocol HomeFlow: AnyObject {
+    func goToCategoriesTab(with itemName: String)
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Home/ViewModels/HomeViewModel.swift:10]: It's strange to use name for items, it's common to use ID

File: Mercury-Store/Mercury-Store/Home/ViewModels/HomeViewModel.swift [hunk 5]
```diff
@@ -6,8 +6,15 @@
...
+    func goToBrandDetails(with brandItem: SmartCollection)
+}
 
 struct HomeViewModel {
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Home/ViewModels/HomeViewModel.swift:14]: Please take care of ordering
// MARK: Properties
// MARK: Lifecycle - Init
// MARK: Methods

File: Mercury-Store/Mercury-Store/Home/Coordinator/HomeCoordinator.swift [hunk 6]
```diff
@@ -7,59 +7,41 @@
...
-        navigationRootViewController?.pushViewController(SearchViewController(coordinator: self), animated: false)
+    init(navigationController : UINavigationController) {
+        self.navigationController = navigationController
     }
     
-    func goToFavouriteScreen() {
-        //TODO: navigate to favourite screen and push the view controller to the stack
+    deinit {
+        print("Deinit home coordinator")
     }
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Home/Coordinator/HomeCoordinator.swift:33]:
```suggestion
```
No debug prints to be merged into main branches

File: Mercury-Store/Mercury-Store/CategoryTabScreen/Views/CategoryMain/Extensions/CategoriesCollectionsExtension.swift [hunk 7]
```diff
@@ -24,6 +24,6 @@ extension CategoryViewController : UICollectionViewDelegate, UICollectionViewDat
         return CGSize(width: collectionViewSize/2, height: 200)
     }
     func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
-        coordinator?.moveTo(flow: .category(.productsScreen), userData: nil)
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/CategoryTabScreen/Views/CategoryMain/Extensions/CategoriesCollectionsExtension.swift:27]: Please add flow or remove documented code entirely

File: Mercury-Store/Mercury-Store/Application/SceneDelegate.swift [hunk 8]
```diff
@@ -10,15 +10,18 @@ import UIKit
...
+    var appCoordinator : ApplicationCoordinator?
 
     func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
         AppAppearance.setupAppearanceForNavigation()
         AppAppearance.setupAppearanceForTabbar()
         guard let windowScene = (scene as? UIWindowScene) else { return }
         
         window = UIWindow(windowScene: windowScene)
-        window?.rootViewController = MainCoordinator().start()
+        let navigationCon = UINavigationController.init()
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Application/SceneDelegate.swift:21]:
```suggestion
        let navigationController = UINavigationController()
```

File: Mercury-Store/Mercury-Store/Support/Coordinator/HomeTabBarCoordinator.swift [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    
+    deinit {
+        print("HomeTabBar Coordinator deinit")
+    }
+
+    
+    func initializeHomeTabBar(){
+        
+        let homeNavigationController = UINavigationController()
+        let homeCoordinator = HomeCoordinator.init(navigationController: homeNavigationController)
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Support/Coordinator/HomeTabBarCoordinator.swift:34]:
```suggestion
        let homeCoordinator = HomeCoordinator(navigationController: homeNavigationController)
```
Screen Shot 2022-06-02 at 10 22 49 PM

File: Mercury-Store/Mercury-Store/Support/Coordinator/HomeTabBarCoordinator.swift [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        cartCoordinator.start()
+        profileCoordinator.start()
+    }
+    
+    init(navigationController : UINavigationController) {
+        self.navigationController = navigationController
+        self.mainTabBar = UITabBarController()
+    }
+    
+}
```
Comment (ahmdmhasn) [inline Mercury-Store/Mercury-Store/Support/Coordinator/HomeTabBarCoordinator.swift:96]:
```suggestion
import Foundation
import UIKit

// MARK: HomeTabBarCoordinator
//
final class HomeTabBarCoordinator: Coordinator {
    weak var parentCoordinator: Coordinator?
    var children: [Coordinator] = []
    var navigationController: UINavigationController
    
    private let mainTabBar: UITabBarController
    
    init(navigationController : UINavigationController) {
        self.navigationController = navigationController
        self.mainTabBar = UITabBarController()
    }
    
    func start() {
        initializeHomeTabBar()
    }
}

// MARK: Private Handlers
//
extension HomeTabBarCoordinator {

    /// Used to make tabs and main coordinators
    ///
    func initializeHomeTabBar() {
        let tabs: [TabContent] = [
            TabContent(title: "Home",
                       image: UIImage(systemName: "homekit"),
                       coordinator: HomeCoordinator(navigationController: UINavigationController())),
            TabContent(title: "Category",
                       image: UIImage(systemName: "square.grid.2x2.fill"),
                       coordinator: CategoryCoordinator(navigationController: UINavigationController())),
            TabContent(title: "Cart",
                       image: UIImage(systemName: "cart.fill"),
                       coordinator: ShoppingCartCoordinator(navigationController: UINavigationController())),
            TabContent(title: "Profile",
                       image: UIImage(systemName: "person.crop.circle.fill"),
                       coordinator: ProfileCoordinator(navigationController: UINavigationController())),
        ]

        tabs.forEach { tab in
            let tabBarItem = UITabBarItem(title: tab.title, image: tab.image)
            tab.coordinator.navigationController.tabBarItem = tabBarItem
            tab.coordinator.parentCoordinator = parentCoordinator
            tab.coordinator.start()
            parentCoordinator?.children.append(tab.coordinator)
        }
        
        mainTabBar.viewControllers = tabs.map(\.coordinator.navigationController)
    }
    
    /// Configure navigation bar appearance
    ///
    func configureNavController() {
        navigationController.pushViewController(mainTabBar, animated: true)
        navigationController.setNavigationBarHidden(true, animated: true)
    }
}

// MARK: Nested Types
//
extension HomeTabBarCoordinator {
    
    struct TabContent {
        let title: String
        let image: UIImage
        let coordinator: Coordinator
    }
}
```

**Review by ahmdmhasn — COMMENTED** [review]
Great effort.. 
Please make sure to keep the changes files as minimum as you can

File: Mercury-Store/Mercury-Store/Home/ViewModels/HomeViewModel.swift (hunk 4 above)
Comment (mahmoudmorsy201) [reply Mercury-Store/Mercury-Store/Home/ViewModels/HomeViewModel.swift:10]: yea but it is a proof of concept on how you can transfer data with the coordinator

File: Mercury-Store/Mercury-Store/Home/Coordinator/HomeCoordinator.swift (hunk 6 above)
Comment (mahmoudmorsy201) [reply Mercury-Store/Mercury-Store/Home/Coordinator/HomeCoordinator.swift:33]: i was testing the life cycle of the coordinator I'm  gonna clear all the print statments

File: Mercury-Store/Mercury-Store/Support/Coordinator/HomeTabBarCoordinator.swift (hunk 10 above)
Comment (mahmoudmorsy201) [reply Mercury-Store/Mercury-Store/Support/Coordinator/HomeTabBarCoordinator.swift:96]: That is better thank you ahmed

File: Mercury-Store/Mercury-Store/ShoppingCart/ViewModels/ShoppingCartViewModel.swift [hunk 11]
```diff
@@ -73,4 +77,14 @@ struct CartViewModel {
...
 
+extension CartViewModel {
+    init(shoppingCartNavigationFlow: ShoppingCartNavigationFlow) {
+        self.shoppingCartNavigationFlow = shoppingCartNavigationFlow
+    }
+}
+
+extension CartViewModel: ShoppingCartNavigationFlow {
+    
+}
```
[the remark this answers is not in the document]
Comment (mahmoudmorsy201) [reply Mercury-Store/Mercury-Store/ShoppingCart/ViewModels/ShoppingCartViewModel.swift:88]: I can't understand totally what you mean 
self.shoppingCartNavigationFlow is a reference to a protocol that will be injected by the coordinator when setting the dependencies you need

**Review by raiaan — APPROVED** [review]

**Pull request merged**